File: frontend/src/types/ship.ts
export interface Engine {
  id: string;
  ship_id: string;
  role: string;
  designation: string;
  manufacturer: string | null;
  model: string | null;
  mcr_kw: number;
  rpm_rated: number | null;
  engine_type: string | null;
  primary_fuel_type: string;
  sfc_g_kwh: number;
  is_dual_fuel: boolean;
  secondary_fuel_type: string | null;
  secondary_sfc_g_kwh: number | null;
  nox_tier: string | null;
}

export interface Ship {
  id: string;
  owner_id: string | null;
  imo_number: string;
  name: string;
  flag_state: string;
  ship_type: string;
  dwt: number;
  gt: number;
  nt: number | null;
  build_date: string;
  delivery_date: string | null;
  reference_speed_kn: number;
  design_draught_m: number;
  fw: number;
  block_coefficient: number | null;
  attained_eedi: number | null;
  required_eedi: number | null;
  eedi_phase: number | null;
  attained_eexi: number | null;
  required_eexi: number | null;
  eexi_compliance_method: string | null;
  epl_limited_power_kw: number | null;
  ice_class: string | null;
  has_cargo_heating: boolean;
  is_shuttle_tanker: boolean;
  is_self_unloading_bc: boolean;
  ism_doc_holder: string | null;
  ops_capable: boolean;
  ops_installed_kw: number | null;
  pooling_group_id: string | null;
  class_society: string | null;
  notes: string | null;
  created_at: string;
  updated_at: string;
  engines: Engine[];
}

export interface EngineCreate {
  role: string;
  designation: string;
  manufacturer?: string | null;
  model?: string | null;
  mcr_kw: number;
  rpm_rated?: number | null;
  engine_type?: string | null;
  primary_fuel_type: string;
  sfc_g_kwh: number;
  is_dual_fuel: boolean;
  secondary_fuel_type?: string | null;
  secondary_sfc_g_kwh?: number | null;
  nox_tier?: string | null;
}

export interface ShipCreate {
  imo_number: string;
  name: string;
  flag_state: string;
  ship_type: string;
  dwt: number;
  gt: number;
  build_date: string;
  reference_speed_kn: number;
  design_draught_m: number;
  engines?: EngineCreate[];
  [key: string]: unknown;
}

export type ShipUpdate = Partial<Omit<ShipCreate, 'imo_number'>>;

export const SHIP_TYPES = [
  { value: 'bulk_carrier', label: 'Bulk Carrier' },
  { value: 'tanker', label: 'Tanker' },
  { value: 'container', label: 'Container Ship' },
  { value: 'gas_carrier', label: 'Gas Carrier' },
  { value: 'general_cargo', label: 'General Cargo' },
  { value: 'refrigerated_cargo', label: 'Refrigerated Cargo' },
  { value: 'combination_carrier', label: 'Combination Carrier' },
  { value: 'lng_carrier', label: 'LNG Carrier' },
  { value: 'cruise_passenger', label: 'Cruise / Passenger' },
  { value: 'roro_cargo', label: 'RoRo Cargo' },
  { value: 'roro_passenger', label: 'RoRo Passenger' },
  { value: 'vehicle_carrier', label: 'Vehicle Carrier' },
] as const;

File: backend/app/schemas/ship.py
from datetime import date, datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict


class EngineBase(BaseModel):
    role: str
    designation: str
    manufacturer: str | None = None
    model: str | None = None
    mcr_kw: float
    rpm_rated: float | None = None
    engine_type: str | None = None
    primary_fuel_type: str
    sfc_g_kwh: float
    is_dual_fuel: bool = False
    secondary_fuel_type: str | None = None
    secondary_sfc_g_kwh: float | None = None
    nox_tier: str | None = None


class EngineCreate(EngineBase):
    pass


class EngineRead(EngineBase):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    ship_id: UUID


class EngineUpdate(BaseModel):
    role: str | None = None
    designation: str | None = None
    manufacturer: str | None = None
    model: str | None = None
    mcr_kw: float | None = None
    rpm_rated: float | None = None
    engine_type: str | None = None
    primary_fuel_type: str | None = None
    sfc_g_kwh: float | None = None
    is_dual_fuel: bool | None = None
    secondary_fuel_type: str | None = None
    secondary_sfc_g_kwh: float | None = None
    nox_tier: str | None = None


class ShipBase(BaseModel):
    imo_number: str
    name: str
    flag_state: str
    ship_type: str
    dwt: float
    gt: float
    nt: float | None = None
    build_date: date
    delivery_date: date | None = None
    reference_speed_kn: float
    design_draught_m: float
    fw: float = 1.0
    block_coefficient: float | None = None
    attained_eedi: float | None = None
    required_eedi: float | None = None
    eedi_phase: int | None = None
    attained_eexi: float | None = None
    required_eexi: float | None = None
    eexi_compliance_method: str | None = None
    epl_limited_power_kw: float | None = None
    ice_class: str | None = None
    has_cargo_heating: bool = False
    is_shuttle_tanker: bool = False
    is_self_unloading_bc: bool = False
    ism_doc_holder: str | None = None
    ops_capable: bool = False
    ops_installed_kw: float | None = None
    pooling_group_id: UUID | None = None
    class_society: str | None = None
    notes: str | None = None


class ShipCreate(ShipBase):
    engines: list[EngineCreate] = []


class ShipRead(ShipBase):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    owner_id: UUID | None = None
    created_at: datetime
    updated_at: datetime
    engines: list[EngineRead] = []


class ShipUpdate(BaseModel):
    name: str | None = None
    flag_state: str | None = None
    ship_type: str | None = None
    build_date: date | None = None
    delivery_date: date | None = None
    dwt: float | None = None
    gt: float | None = None
    nt: float | None = None
    reference_speed_kn: float | None = None
    design_draught_m: float | None = None
    fw: float | None = None
    block_coefficient: float | None = None
    attained_eedi: float | None = None
    required_eedi: float | None = None
    eedi_phase: int | None = None
    attained_eexi: float | None = None
    required_eexi: float | None = None
    eexi_compliance_method: str | None = None
    epl_limited_power_kw: float | None = None
    ice_class: str | None = None
    has_cargo_heating: bool | None = None
    is_shuttle_tanker: bool | None = None
    is_self_unloading_bc: bool | None = None
    ism_doc_holder: str | None = None
    ops_capable: bool | None = None
    ops_installed_kw: float | None = None
    pooling_group_id: UUID | None = None
    class_society: str | None = None
    notes: str | None = None
    engines: list[EngineCreate] | None = None


class ShipList(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    imo_number: str
    name: str
    flag_state: str
    ship_type: str
    dwt: float
    gt: float
